File: ExtremeStudio/Functions Forms/ESPluginsForm.vb
Imports System.IO
Imports System.Net
Imports System.Xml

Public Class EsPluginsForm
    Dim _plugins As New List(Of EsPluginData)
    Dim _isThereInteret As Boolean = True

    Private Sub RefreshList()
        PluginList.Items.Clear()
        For Each plug In _plugins
            PluginList.Items.Add(plug.PluginName)
        Next
    End Sub

    Private Function IsPluginInstalled(plug as EsPluginData)
        Dim file As String = Path.GetFileName(plug.PluginPath)
        If My.Computer.FileSystem.FileExists(Mainform.ApplicationFiles + "/plugins/" + file)
            Return True
        Else
            Return False
        End If
    End Function
    Private Function IsPluginDifferent(plug As EsPluginData)
        Dim file As String = Path.GetFileName(plug.PluginPath)
        Dim oldHash As String = ExtremeCore.GetFileHash(Mainform.ApplicationFiles + "/plugins/" + file)
        If oldHash = plug.PluginSha
            Return False
        Else
            Return True
        End If
    End Function

    Private Sub EsPluginsForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        _plugins.Clear()

        If ExtremeCore.isNetAvailable() Then 'Download latest from internet if there is internet.
            Dim webClient As New WebClient
            Dim xmlFile As New XmlDocument()
            Dim fileText As String = webClient.DownloadString("http://johnymac.github.io/esfiles/esplugins.xml")
            xmlFile.LoadXml(fileText)

            For Each cs As XmlNode In xmlFile.SelectNodes("xml/plugin")
                Dim plug As New EsPluginData
                plug.PluginName = cs("name").InnerText.Trim
                plug.PluginDesc = cs("desc").InnerText.Trim
                plug.PluginPath = cs("download").InnerText.Trim
                plug.PluginSha = cs("hash").InnerText.Trim
                _plugins.Add(plug)
            Next
        Else
            _isThereInteret = False
            For Each file In Directory.GetFiles(MainForm.ApplicationFiles + "/plugins", "*.dll")
                _plugins.Add(New EsPluginData() With { .PluginName = Path.GetFileName(file) })
            Next
        End If

        RefreshList()
    End Sub

    Private Sub PluginList_Click(sender As Object, e As EventArgs) Handles PluginList.Click
            If PluginList.SelectedIndex <> - 1 Then
                If _isThereInteret Then
                    For Each plug In _plugins
                        If plug.PluginName = PluginList.SelectedItem
                            PluginNameText.Text = plug.PluginName
                            PluginDescText.Text = plug.PluginDesc
                            If IsPluginInstalled(plug)
                                installBtn.Text = "ReInstall Plugin"
                                installBtn.Enabled = True
                                deleteBtn.Visible = True
                                If IsPluginDifferent(plug)
                                    updateLabel.Visible = True
                                End If
                            Else
                                installBtn.Text = "Install Plugin"
                                installBtn.Enabled = True
                                deleteBtn.Visible = False
                                updateLabel.Visible = False
                            End If
                        End If
                    Next
                Else 
                    installBtn.Text = ""
                    installBtn.Enabled = False
                    deleteBtn.Visible = True
                    updateLabel.Visible = False
                End If
            Else
                installBtn.Text = ""
                installBtn.Enabled = False
                deleteBtn.Visible = False
                updateLabel.Visible = False
            End If

    End Sub

    Private Sub installBtn_Click(sender As Object, e As EventArgs) Handles installBtn.Click
        If _isThereInteret
            If PluginList.SelectedIndex <> -1
                Dim plugPath As String = Nothing
                For Each plug In _plugins
                    If plug.PluginName = PluginList.SelectedItem Then
                        plugPath = Mainform.ApplicationFiles + "/plugins/" + Path.GetFileName(plug.PluginPath)
                        If File.Exists(plugPath) Then File.Delete(plugPath)
                        Dim web As New WebClient
                        web.DownloadFile(plug.PluginPath, plugPath)
                        EsPluginsForm_Load(Me, Nothing)
                        PluginList_Click(Me, Nothing)
                        Exit For
                    End If
                Next
            End If
        End If
    End Sub

    Private Sub deleteBtn_Click(sender As Object, e As EventArgs) Handles deleteBtn.Click
        If _isThereInteret
            If PluginList.SelectedIndex <> -1
                Dim plugPath As String = Nothing
                For Each plug In _plugins
                    If plug.PluginName = PluginList.SelectedItem Then
                        plugPath = Mainform.ApplicationFiles + "/plugins/" + Path.GetFileName(plug.PluginPath)
                        If File.Exists(plugPath) Then File.Delete(plugPath)
                        EsPluginsForm_Load(Me, Nothing)
                        PluginList_Click(Me, Nothing)
                        Exit For
                    End If
                Next
            End If
        Else
            File.Delete(Mainform.ApplicationFiles + "/plugins/" + PluginList.SelectedItem)
            EsPluginsForm_Load(Me, Nothing)
            PluginList_Click(Me, Nothing)
        End If
    End Sub
End Class

Public Class EsPluginData
    Public Property PluginName As String
    Public Property PluginDesc As String
    Public Property PluginPath As String
    Public Property PluginSha As String
End Class

File: ExtremeStudio/FunctionsForms/ESPluginsForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml;
using Resources;

namespace ExtremeStudio.FunctionsForms
{
    public partial class ESPluginsForm : Form
    {
        public ESPluginsForm()
        {
            InitializeComponent();
        }

        List<EsPluginData> _plugins = new List<EsPluginData>();
        bool _isThereInteret = true;

        private void RefreshList()
        {
            PluginList.Items.Clear();
            foreach (var plug in _plugins)
            {
                PluginList.Items.Add(plug.PluginName);
            }
        }

        private dynamic IsPluginInstalled(EsPluginData plug)
        {
            string file = Path.GetFileName(plug.PluginPath);
            if (File.Exists(
                Application.StartupPath + "/plugins/" + file))
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        private dynamic IsPluginDifferent(EsPluginData plug)
        {
            string file = Path.GetFileName(plug.PluginPath);
            string oldHash =
                System.Convert.ToString(ExtremeCore.Classes.GeneralFunctions.GetFileHash(Application.StartupPath + "/plugins/" + file));
            if (oldHash == plug.PluginSha)
            {
                return false;
            }
            else
            {
                return true;
            }
        }

        private void EsPluginsForm_Load(object sender, EventArgs e)
        {
            _plugins.Clear();

            if (ExtremeCore.Classes.GeneralFunctions.IsNetAvailable()) //Download latest from internet if there is internet.
            {
                WebClient webClient = new WebClient();
                XmlDocument xmlFile = new XmlDocument();
                string fileText =
                    System.Convert.ToString(
                        webClient.DownloadString("http://ahmad45123.github.io/esfiles/esplugins.xml"));
                xmlFile.LoadXml(fileText);

                foreach (XmlNode cs in xmlFile.SelectNodes("xml/plugin"))
                {
                    EsPluginData plug = new EsPluginData();
                    plug.PluginName = System.Convert.ToString(cs["name"].InnerText.Trim());
                    plug.PluginDesc = System.Convert.ToString(cs["desc"].InnerText.Trim());
                    plug.PluginPath = System.Convert.ToString(cs["download"].InnerText.Trim());
                    plug.PluginSha = System.Convert.ToString(cs["hash"].InnerText.Trim());
                    _plugins.Add(plug);
                }
            }
            else
            {
                _isThereInteret = false;
                foreach (var file in Directory.GetFiles(Application.StartupPath + "/plugins", "*.dll"))
                {
                    _plugins.Add(new EsPluginData() {PluginName = Path.GetFileName(System.Convert.ToString(file))});
                }
            }

            RefreshList();
        }

        private void PluginList_Click(object sender, EventArgs e)
        {
            if (PluginList.SelectedIndex != -1)
            {
                if (_isThereInteret)
                {
                    foreach (var plug in _plugins)
                    {
                        if (plug.PluginName == (string)PluginList.SelectedItem)
                        {
                            PluginNameText.Text = plug.PluginName;
                            PluginDescText.Text = plug.PluginDesc;
                            if (IsPluginInstalled(plug))
                            {
                                installBtn.Text = translations.EsPluginsForm_PluginList_Click_ReInstallPlugin;
                                installBtn.Enabled = true;
                                deleteBtn.Visible = true;
                                if (IsPluginDifferent(plug))
                                {
                                    updateLabel.Visible = true;
                                }
                            }
                            else
                            {
                                installBtn.Text = translations.EsPluginsForm_PluginList_Click_InstallPlugin;
                                installBtn.Enabled = true;
                                deleteBtn.Visible = false;
                                updateLabel.Visible = false;
                            }
                        }
                    }
                }
                else
                {
                    installBtn.Text = "";
                    installBtn.Enabled = false;
                    deleteBtn.Visible = true;
                    updateLabel.Visible = false;
                }
            }
            else
            {
                installBtn.Text = "";
                installBtn.Enabled = false;
                deleteBtn.Visible = false;
                updateLabel.Visible = false;
            }

        }

        private void installBtn_Click(object sender, EventArgs e)
        {
            if (_isThereInteret)
            {
                if (PluginList.SelectedIndex != -1)
                {
                    foreach (var plug in _plugins)
                    {
                        if (plug.PluginName == (string) PluginList.SelectedItem)
                        {
                            var plugPath = Application.StartupPath + "/plugins/" +
                                              Path.GetFileName(System.Convert.ToString(plug.PluginPath));
                            if (File.Exists(plugPath))
                            {
                                File.Delete(plugPath);
                            }

                            WebClient web = new WebClient();
                            web.DownloadFile(plug.PluginPath, plugPath);
                            EsPluginsForm_Load(this, null);
                            PluginList_Click(this, null);
                            break;
                        }
                    }
                }
            }
        }

        private void deleteBtn_Click(object sender, EventArgs e)
        {
            if (_isThereInteret)
            {
                if (PluginList.SelectedIndex != -1)
                {
                    string plugPath = null;
                    foreach (var plug in _plugins)
                    {
                        if (plug.PluginName == (string)PluginList.SelectedItem)
                        {
                            plugPath = Application.StartupPath + "/plugins/" +
                                       Path.GetFileName(System.Convert.ToString(plug.PluginPath));
                            if (File.Exists(plugPath))
                            {
                                File.Delete(plugPath);
                            }

                            EsPluginsForm_Load(this, null);
                            PluginList_Click(this, null);
                            break;
                        }
                    }
                }
            }
            else
            {
                File.Delete(Application.StartupPath + "/plugins/" + PluginList.SelectedItem);
                EsPluginsForm_Load(this, null);
                PluginList_Click(this, null);
            }
        }
    }

    public class EsPluginData
    {
        public string PluginName {get; set;}
        public string PluginDesc {get; set;}
        public string PluginPath {get; set;}
        public string PluginSha {get; set;}
    }
}